substrate/frame/revive/src/evm/block_hash.rs

```diff
@@ -263,6 +99,504 @@ impl Block {
+			0 => {
+				let mut value = HashBuilderValue::new();
+				value.set_bytes_owned(serialized.builder_value);
+				value
+			},
+			1 => {
+				use alloy_core::primitives::B256;
+
+				let buffer: B256 = serialized.builder_value[..]
+					.try_into()
+					.expect("The buffer was serialized properly; qed");
+				let value_ref = HashBuilderValueRef::Hash(&buffer);
+
+				let mut value = HashBuilderValue::new();
+				value.set_from_ref(value_ref);
+				value
+			},
+			_ => panic!("Value type was serialized properly; qed"),
+		};
+
+		let hash_builder = HashBuilder {
+			key: Nibbles::from_nibbles(serialized.key),
+			value,
+			stack: serialized
+				.stack
+				.into_iter()
+				.map(|raw| RlpNode::from_raw(&raw).expect("RlpNode was encoded properly; qed"))
+				.collect(),
+			state_masks: serialized
+				.state_masks
+				.into_iter()
+				.map(|mask| TrieMask::new(mask))
+				.collect(),
+			tree_masks: serialized.tree_masks.into_iter().map(|mask| TrieMask::new(mask)).collect(),
+			hash_masks: serialized.hash_masks.into_iter().map(|mask| TrieMask::new(mask)).collect(),
+			stored_in_database: serialized.stored_in_database,
+			updated_branch_nodes: None,
+			proof_retainer: None,
+			rlp_buf: serialized.rlp_buf,
+		};
+
+		IncrementalHashBuilder {
+			hash_builder,
+			index: serialized.index,
+			first_value: serialized.first_value,
+		}
+	}
+
+	/// Converts the builder into an intermediate representation.
+	pub fn to_ir(self) -> IncrementalHashBuilderIR {
+		use alloy_consensus::private::alloy_trie::hash_builder::HashBuilderValueRef;
+
+		IncrementalHashBuilderIR {
+			key: self.hash_builder.key.to_vec(),
+			value_type: match self.hash_builder.value.as_ref() {
+				HashBuilderValueRef::Bytes(_) => 0,
+				HashBuilderValueRef::Hash(_) => 1,
+			},
+			builder_value: self.hash_builder.value.as_slice().to_vec(),
+			stack: self.hash_builder.stack.into_iter().map(|n| n.as_slice().to_vec()).collect(),
+
+			state_masks: self.hash_builder.state_masks.into_iter().map(|mask| mask.get()).collect(),
+			tree_masks: self.hash_builder.tree_masks.into_iter().map(|mask| mask.get()).collect(),
+			hash_masks: self.hash_builder.hash_masks.into_iter().map(|mask| mask.get()).collect(),
+
+			stored_in_database: self.hash_builder.stored_in_database,
+			rlp_buf: self.hash_builder.rlp_buf,
+			index: self.index,
+			first_value: self.first_value,
+		}
+	}
+
+	/// Add a new value to the hash builder.
+	pub fn add_value(&mut self, value: Vec<u8>) {
+		if self.index == 0x7f {
+			// Pushing the previous item since we are expecting the index
+			// to be index + 1 in the sorted order.
+			if let Some(encoded_value) = self.first_value.take() {
+				let rlp_index = alloy_rlp::encode_fixed_size(&0usize);
+
+				self.hash_builder.add_leaf(Nibbles::unpack(&rlp_index), &encoded_value);
+			}
+		}
+
+		let rlp_index = alloy_rlp::encode_fixed_size(&self.index);
+		self.hash_builder.add_leaf(Nibbles::unpack(&rlp_index), &value);
```

Comment:
We are inserting in this order:  `([128], first_value)`, `([127], 127th_value)`
We want to insert in this order: `([127], 127th_value)`, `([128], first_value)`

```suggestion
		let rlp_index = alloy_rlp::encode_fixed_size(&self.index);
		self.hash_builder.add_leaf(Nibbles::unpack(&rlp_index), &value);

		if self.index == 0x7f {
			// Pushing the previous item since we are expecting the index
			// to be index + 1 in the sorted order.
			if let Some(encoded_value) = self.first_value.take() {
				let rlp_index = alloy_rlp::encode_fixed_size(&0usize);

				self.hash_builder.add_leaf(Nibbles::unpack(&rlp_index), &encoded_value);
			}
		}
```